Reform BoolLike

garyb opened the pull request:
Resolves #907.

garyb commented:
Disregard that, I gave `ComplementedLattice` the "relatively complemented" law instead.

paf31 commented:
Looks good to me. A couple of things:

- `BoolLike` can be renamed if we like. `Boolean`?
- Is it worth introducing `Ord <= Bounded` and `BoundedLattice`?
- It might be good to use `meet` and `join`, with aliases for `(&&)` and `(||)`?
- Does the optimizer need changes?

garyb commented:
1. I did wonder about renaming, `BooleanAlgebra` was my first choice, but there are one or two things that broke in `test-everything` then. I think it may have been `lens` related again though.

2. So we'd have `Ord <= Lattice`, `Ord <= Bounded`, `Lattice, Bounded <= BoundedLattice`? That's fine by me, I mashed `Lattice` and `BoundedLattice` together to cut down on the number of classes required, making the (perhaps erroneous) assumption that most lattices we'd have would be finite. Would `Bounded` have any laws of its own?

3. Good point with the operators. That'll save pain when we do the aliasing stuff. :+1:

4. Yep, it does, I meant to mention that in the opening too.

paf31 commented:
> Would Bounded have any laws of its own?

```purescript
bottom <= x <= top
```

is the only thing I can think of.

garyb commented:
Updated with `Bounded` etc.

I also put the complemented and relatively complemented laws back in. I'd appreciate an opinion from someone who actually knows what they're talking about in regards to that! Can reword the docs to avoid the quantification symbols, but I just wanted to make it clear what I thought they are, whether they're right, whether we *should* make `ComplementedLattice` relatively complemented, etc

I have a slight concern about using `join` after thinking about this, as we also have `Control.Monad.join` and I think there may be a `String` or `Array` `join` too.

paf31 commented:
Good point. `sup` and `lub`?

I'll read a bit about complemented lattices.

garyb commented:
Wouldn't it be `sup` and `inf` or `lub` and `glb`?

Can't say I'm crazy about any of those names either ;)

paf31 commented:
Yes, sorry :)

garyb commented:
Regarding the laws, I think it is right the way I have it now:

- It is not the case that every bounded lattice must be relatively complemented
- A distributive lattice is only complemented iff it is relatively complemented

But stating the latter law may be redundant, as I assume it's not actually possible to define a lattice that satisfies both **distributivity** and **complemented** unless it is relatively complemented.

garyb commented:
I think I'd prefer to rename the `Monad` `join` that this one actually, this seems the more generic operation, so perhaps the other could become `joinM` - having _operations_ called `inf` and `sup` seems a little confusing.

I've updated `Optimizer` to inline `Boolean` operations again, and added more content to the docs now.

paf31 commented on prelude/README.md:
````diff
@@ -150,7 +150,7 @@ products # groupBy productCategory # filter isInStock # length
 
 ```purescript
-length (groupBy productCategory (filter isInStock (products)))
+length (filter isInStock (groupBy productCategory (products)))
````
Actually, I think the other one was wrong :) The idea was that `isInStock` applies to a product.

garyb replied:
Oh, I didn't actually do this, it must have been a change made in `prelude.purs` at some point and then the docs hadn't been regenerated...

paf31 commented on prelude/README.md:
```diff
@@ -782,49 +768,169 @@ instance ordArray :: (Ord a) => Ord [a]
-#### `bitsNumber`
+Instances should satisfy the following law in addition to the `Ord` laws:
+
+- Ordering: `bottom <= top`
```
It's stronger: `forall x. bottom <= x <= top`

paf31 commented on prelude/README.md:
```diff
@@ -782,49 +768,169 @@ instance ordArray :: (Ord a) => Ord [a]
+`BoundedLattice` laws:
+
+- Complemented:
+  - For every `a` in interval `[c, d]` there should be a `b` such that `a || b = top`
```
What are `c` and `d` here?

garyb replied:
`[c, d]` is just "any interval in the poset", I guess since `c` and `d` aren't used here (unlike the relatively complemented law) we could just say it that way instead?

paf31 replied:
Can we just say `not a || a = top` and `not a && a = bottom`?

paf31 replied:
What I'm saying is that we don't require _some_ element to have that property, it's `not a` in particular that has it. I guess that's implied though. Maybe I'm missing the point of `[c, d]`, but if you pick `c = d = a = bottom`, then it's not true, right? Do we also need `not bottom = top` and vice versa?

garyb replied:
I got this wrong actually, the interval stuff is only relevant to relatively complemented lattices - the laws according to wikipedia at least are that given an `a` there is some `b` where `a || b = bottom` and `a && b == top`.

If we can use `¬a` instead of `b` then great, is that definitely right?

paf31 replied:
I think so, yes.

paf31 commented on prelude/README.md:
````diff
@@ -869,6 +975,28 @@ instance semigroupArr :: (Semigroup s') => Semigroup (s -> s')
+#### `Bits`
+
+``` purescript
+class Bits b where
````
Just this one left to fix I guess :)

paf31 commented:
Looks good. Just see my comments, and also we still need to decide on `join` vs `sup` vs whatever. I don't feel too strongly against `sup` and `inf`, since I guess people using them abstractly will probably be familiar with the terms. Actually, I quite like `sup`/`inf`, because the name suggests "bigger"/"smaller", whereas `join`/`meet` don't IMO.

garyb commented:
Alright, there we go.

paf31 commented:
Looks good to me. @jdegoes @joneshf Any comments?

paf31 commented:
The coverage decrease must be the new optimizer rules. Any chance you could add a quick test for those please?

garyb commented:
Eh?

paf31 commented:
Yeah, something's not right there :)

garyb commented:
I couldn't quite figure out from Coveralls what it though was being missed either - I can see which files it says coverage has decreased in (all of which look suspect!) but not the exact lines.

joneshf commented on prelude/prelude.purs:
```diff
@@ -702,6 +706,167 @@ module Prelude
+  -- |
+  -- | Instances should satisfy the following laws in addition to the `Ord`
+  -- | laws:
+  -- |
```
It might make sense to write the laws with `sup` and `inf`.

garyb replied:
Yeah, I wasn't sure which way to go with that, but I figured it would be easier to read with the operators. That's also why I listed their equivalents in the comment above, even though they're not part of the class definition itself.

joneshf commented on prelude/prelude.purs:
```diff
@@ -702,6 +706,167 @@ module Prelude
+  -- |
+  -- | - Identity:
+  -- |   - `a || bottom = a`
+  -- |   - `a && top = a`
```
Should Annihiliation be here?

```
a || top = top
a && bottom = bottom
```

garyb replied:
Yep, that makes sense.

jdegoes commented:
At a high-level, looks really good to me. Gonna clean up the "boolean" hierarchy quite a lot! :+1:

joneshf commented on prelude/prelude.purs:
```diff
@@ -702,6 +706,167 @@ module Prelude
+  -- |
+  -- | - Complemented:
+  -- |   - `not a || a == top`
+  -- |   - `not a && a == bottom`
```
Do we need to mention double negation?
```
not >>> not = id
```

joneshf replied:
Also, what about DeMorgan's?

```
not a || not b == not (a && b)
not a && not b == not (a || b)
```

joneshf replied:
Do we need to say something about `top` and `bottom`?

```
not top == bottom
not bottom == top
```

garyb replied:
`not >>> not = id` makes sense, but is that implied by the fact it's uniquely complemented anyway? I guess it does no harm to list it. I guess the description should mention that it's uniquely complemented also, even though from the type signature doesn't admit anything else.

garyb replied:
It has to be complemented _and_ distributive for DeMorgan's doesn't it? So would only apply to `BooleanAlgebra` in this hierarchy, and would be implied by the other laws anyway (I think).

joneshf commented on prelude/prelude.purs:
```diff
@@ -702,6 +706,167 @@ module Prelude
+  -- | algebras, also known as Boolean lattices.
+  -- |
+  -- | Instances should satisfy the `ComplementedLattice` and
+  -- | `DistributiveLattice` laws.
```
Actually, you probably need distributativity and negation to prove DeMorgan's. Though it may be derivable free from the other laws?

joneshf commented:
Looks good.

It doesn't have to be this PR, but I think we might be able to add some instances for `Number` and `Unit`.

`-Infinity`/`Infinity` for `Bounded`, https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Infinity.

And `Unit` should satisfy all of them I believe.

garyb commented:
Good call, I intend to add a bunch of `Unit` instances after this is merged actually.

garyb commented:
Ok, updated with a few extra laws, and squashed the commits.

garyb commented:
Should we just go ahead and remove `Bits` while we're at it, and move it to `purescript-integers`?

jdegoes commented:
:+1:

paf31 commented:
I was thinking that too. It'll break the optimizations, but that's probably not the end of the world.

Fresheyeball commented:
Sooo awesome. :+1:
